src/nupic/encoders/CoordinateEncoder.hpp

```diff
@@ -0,0 +1,369 @@
+*/
+#ifndef NTA_coordinateencoder_HPP
+#define NTA_coordinateencoder_HPP
+
+#include <algorithm> //min
+#include <cstring>
+#include <iostream>
+#include <iterator>
+#include <string>
+#include <vector>
+#include <tuple> 
+#include <algorithm>
+
+#include <nupic/encoders/Encoder.hpp>
+#include <nupic/encoders/Utils.hpp>
+#include <nupic/encoders/md5.hpp> //TODO possible update to https://tls.mbed.org/md5-source-code , but it's C
+#include <nupic/utils/Random.hpp>
+
+#define _USE_BIG_INTEGER__ //Always ise BIG_INTEGER, the other way does not work
+#ifdef _USE_BIG_INTEGER__
+#include <nupic/encoders/bigInt.h>
+#endif
+
+using namespace std;
+using namespace nupic;
+using coord_t = vector<int>;
+
+namespace encoders {
+
+////// INTERFACE 
+        class CoordinateOrder {     // An interface class
+        public:
+                CoordinateOrder() {};
+                virtual ~CoordinateOrder() {};   // destructor, use it to call destructor of the inherit classes
+
+                /**
+                * Returns the order for a coordinate.
+                *
+                * @param coordinate             coordinate array
+                *
+                * @return       A value in the interval [0, 1), representing the
+                *          order of the coordinate
+                */
+                virtual double orderForCoordinate(coord_t coordinate) const =0;
+
+        }; // end class CoordinateOrder
+
+////////
+
+	/**
+	* Given a coordinate in an N-dimensional space, and a radius around
+	* that coordinate, the Coordinate Encoder returns an SDR representation
+	* of that position.
+	* The Coordinate Encoder uses an N-dimensional integer coordinate space.
+	* For example, a valid coordinate in this space is (150, -49, 58), whereas
+	* an invalid coordinate would be (55.4, -5, 85.8475).
+	* It uses the following algorithm:
+	* 1. Find all the coordinates around the input coordinate, within the
+	* specified radius.
+	* 2. For each coordinate, use a uniform hash function to
+	* deterministically map it to a real number between 0 and 1. This is the
+	* "order" of the coordinate.
+	* 3. Of these coordinates, pick the top W by order, where W is the
+	* number of active bits desired in the SDR.
+	* 4. For each of these W coordinates, use a uniform hash function to
+	* deterministically map it to one of the bits in the SDR. Make this bit active.
+	* 5. This results in a final SDR with exactly W bits active
+	* (barring chance hash collisions).
+	* 
+	*/
+	template<typename T >
+	class CoordinateEncoder :  public Encoder<T>, public CoordinateOrder
+	{
+	public:
+		/***************** constructors and destructor *****************/
+
+		/**
+		* Package private to encourage construction using the Builder Pattern
+		* but still allow inheritance.
+		*/
+		CoordinateEncoder(int w =21, int n =1024, double radius=1.0, bool retinaMode=false, vector<UInt> DIMs_nD=vector<UInt>{}); //TODO separate Retina to a standalone constructor?
```

Comment:
CoordinateEncoder can probably be replaced by SDR and its `setFromCoordinates()` functionality. 

CoordEncoder is: 

> * Given a coordinate in an N-dimensional space, and a radius around
	* that coordinate, the Coordinate Encoder returns an SDR representation
	* of that position.
	* The Coordinate Encoder uses an N-dimensional integer coordinate space.
	* For example, a valid coordinate in this space is (150, -49, 58), whereas
	* an invalid coordinate would be (55.4, -5, 85.8475).
	* It uses the following algorithm:
	* 1. Find all the coordinates around the input coordinate, within the
	* specified radius.
	* 2. For each coordinate, use a uniform hash function to
	* deterministically map it to a real number between 0 and 1. This is the
	* "order" of the coordinate.
	* 3. Of these coordinates, pick the top W by order, where W is the
	* number of active bits desired in the SDR.
	* 4. For each of these W coordinates, use a uniform hash function to
	* deterministically map it to one of the bits in the SDR. Make this bit active.
	* 5. This results in a final SDR with exactly W bits active
	* (barring chance hash collisions).


Before doing that, we should consider some differences:
- coord encoder can have negative integer coordinate indices (SDR can have only natural, but that should not be a big limitation) 
- keep/replicate coord encoder's tests

Reply:
The Coordinate Encoder & `SDR::setCoordinates` do very different things.

SDR::setCoordinates sets the literal value of an SDR.

Coordinate Encoder maps a location in space to an SDR.  The resulting SDR has all of the expected properties of encoded SDRs: constant sparsity, semantic similarity, uniqueness.

Reply:
ok, my argument is not valid, and CoordinateEncoder then has its meaning.